src/lib/coreaudio/CoreAudioStream.cxx

```diff
@@ -0,0 +1,293 @@
+	if (AudioObjectRemovePropertyListener(stream_id, &aopa, HardwareStreamListener, this) != noErr)
+		FormatDebug(macos_output_domain, "Couldn't remove property listener for Core Audio stream.");
+
+	// Revert any format changes we made
+	if (restore && original_virtual_fmt.mFormatID && stream_id)
+	{
+		FormatDebug(macos_output_domain, "Restoring original virtual format for stream 0x%04x. (%s)",
+					(uint)stream_id, StreamDescriptionToString(original_virtual_fmt, format_string));
+		AudioStreamBasicDescription setFormat = original_virtual_fmt;
+		SetVirtualFormat(&setFormat);
+	}
+	if (restore && original_physical_fmt.mFormatID && stream_id)
+	{
+		FormatDebug(macos_output_domain, "Restoring original physical format for stream 0x%04x. (%s)",
+					(uint)stream_id, StreamDescriptionToString(original_physical_fmt, format_string));
+		AudioStreamBasicDescription setFormat = original_physical_fmt;
+		SetPhysicalFormat(&setFormat);
+	}
+
+	original_virtual_fmt.mFormatID  = 0;
+	original_physical_fmt.mFormatID = 0;
+	FormatDebug(macos_output_domain, "Closed stream 0x%04x.", (uint)stream_id);
+	stream_id = 0;
+}
+
+bool CoreAudioStream::GetStartingChannelInDevice(AudioStreamID id, UInt32 &startingChannel)
+{
+	if (!id)
+		return 0;
+
+	UInt32 i_param_size = sizeof(UInt32);
+	UInt32 i_param;
+	startingChannel = 0;
+	bool ret = false;
+
+	AudioObjectPropertyAddress aopa;
+	aopa.mScope    = kAudioObjectPropertyScopeGlobal;
+	aopa.mElement  = kAudioObjectPropertyElementMaster;
+	aopa.mSelector = kAudioStreamPropertyStartingChannel;
+
+	OSStatus status = AudioObjectGetPropertyData(id, &aopa, 0, NULL, &i_param_size, &i_param);
+	if (status == noErr)
+	{
+		startingChannel = i_param;
+		ret = true;
+	}
+	return ret;
+}
+
+bool CoreAudioStream::GetVirtualFormat(AudioStreamBasicDescription* desc)
+{
+	if (!desc || !stream_id)
+		return false;
+
+	UInt32 size = sizeof(AudioStreamBasicDescription);
+
+	AudioObjectPropertyAddress aopa;
+	aopa.mScope    = kAudioObjectPropertyScopeGlobal;
+	aopa.mElement  = kAudioObjectPropertyElementMaster;
+	aopa.mSelector = kAudioStreamPropertyVirtualFormat;
+	OSStatus ret = AudioObjectGetPropertyDataSize(stream_id, &aopa, 0, NULL, &size);
+	if (ret)
+		return false;
+
+	ret = AudioObjectGetPropertyData(stream_id, &aopa, 0, NULL, &size, desc);
+	if (ret)
+		return false;
+	return true;
+}
+
+bool CoreAudioStream::SetVirtualFormat(AudioStreamBasicDescription* desc)
+{
+	if (!desc || !stream_id)
+		return false;
+
+	std::string format_string;
+
+	AudioObjectPropertyAddress aopa;
+	aopa.mScope    = kAudioObjectPropertyScopeGlobal;
+	aopa.mElement  = kAudioObjectPropertyElementMaster;
+	aopa.mSelector = kAudioStreamPropertyVirtualFormat;
+
+	UInt32 property_size = sizeof(AudioStreamBasicDescription);
+	OSStatus ret = AudioObjectSetPropertyData(stream_id, &aopa, 0, NULL, property_size, desc);
+	if (ret)
+	{
+		FormatError(macos_output_domain, "Unable to set virtual format for stream 0x%04x. Error = %s",
```

Comment:
Throw exception!!

Reply:
Changed this to void function throwing exception on error